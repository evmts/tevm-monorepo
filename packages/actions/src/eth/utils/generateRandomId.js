/**
 * @returns {import("@tevm/utils").Hex}
 */
export const generateRandomId = () => {
	return `0x${Array.from(crypto.getRandomValues(new Uint8Array(16)))
		.map((b) => b.toString(16).padStart(2, '0'))
		.join('')}`
}
