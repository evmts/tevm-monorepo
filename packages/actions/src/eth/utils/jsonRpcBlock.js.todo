import { bytesToHex } from "viem"

const ethjsTxToTevmTx = (tx) => {

}

/**
 * Transoforms an ethjs block to a tevm Block
 * @param {import("@ethereumjs/block").Block} block
 * @param {boolean} includeTransactions
 * @returns {Promise<import("@tevm/actions-types").Block>}
 */
const ethjsToTevmBlock = async (
  block,
  includeTransactions,
) => {
  const transactions = block.transactions.map((tx, txIndex) =>
    includeTransactions ? ethjsTxToTevmTx(tx, block, txIndex) : bytesToHex(tx.hash())
  )
  const withdrawalsAttr =
    block.header.withdrawalsRoot !== undefined
      ? {
        withdrawalsRoot: bytesToHex(block.header.withdrawalsRoot),
        withdrawals: block.toJSON().withdrawals,
      }
      : {}
  const td = await chain.getTd(block.hash(), block.header.number)
  return {
    coinbase: block.toJSON().header?.coinbase
    number: block.toJSON().header.number,
    hash: bytesToHex(block.hash()),
    parentHash: block.toJSON().header.parentHash,
    mixHash: block.toJSON().header.mixHash,
    nonce: block.toJSON().header.nonce,
    sha3Uncles: block.toJSON().header.uncleHash,
    logsBloom: block.toJSON().header.logsBloom,
    transactionsRoot: block.toJSON().header.transactionsTrie,
    stateRoot: block.toJSON().header.stateRoot,
    receiptsRoot: block.toJSON().header.receiptTrie,
    miner: block.toJSON().header.coinbase,
    difficulty: block.toJSON().header.difficulty,
    totalDifficulty: bigIntToHex(td),
    extraData: block.toJSON().header.extraData,
    size: intToHex(utf8ToBytes(JSON.stringify(block.toJSON())).byteLength),
    gasLimit: block.toJSON().header.gasLimit,
    gasUsed: block.toJSON().header.gasUsed,
    timestamp: block.toJSON().header.timestamp,
    transactions,
    uncles: block.uncleHeaders.map((uh) => bytesToHex(uh.hash())),
    baseFeePerGas: block.toJSON().header.baseFeePerGas,
    ...withdrawalsAttr,
    blobGasUsed: block.toJSON().header.blobGasUsed,
    excessBlobGas: block.toJSON().header.excessBlobGas,
    parentBeaconBlockRoot: block.toJSON().header.parentBeaconBlockRoot,
  }
}
