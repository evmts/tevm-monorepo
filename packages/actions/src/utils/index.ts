export * from './SerializeToJson.js'
export * from './blockToJsonRpcBlock.js'
export * from './generateRandomId.js'
export * from './parseBlockTag.js'
export * from './txToJSONRPCTx.js'
