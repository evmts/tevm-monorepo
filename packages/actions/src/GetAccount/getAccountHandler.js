import { createAddress } from '@tevm/address'
import { AccountNotFoundError, ForkError, InternalError } from '@tevm/errors'
import { bytesToHex } from '@tevm/utils'
import { cloneVmWithBlockTag } from '../Call/cloneVmWithBlock.js'
import { getPendingClient } from '../internal/getPendingClient.js'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { validateGetAccountParams } from './validateGetAccountParams.js'

/**
 * Creates an GetAccountHandler for handling account params with Ethereumjs VM
 * @param {import("@tevm/node").TevmNode} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('../GetAccount/GetAccountHandlerType.js').GetAccountHandler}
 */
export const getAccountHandler =
	(client, options = {}) =>
	async ({ throwOnFail = options.throwOnFail ?? true, ...params }) => {
		const vm = await client.getVm()
		/**
		 * @type {Array<import('../GetAccount/TevmGetAccountError.js').TevmGetAccountError>}
		 */
		const errors = validateGetAccountParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, {
				errors,
				address: params.address,
				balance: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				storageRoot: '0x',
				nonce: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				deployedBytecode: '0x',
				/**
				 * @type {`0x${string}`}
				 */
				codeHash: '0x',
				isContract: false,
				isEmpty: true,
			})
		}

		const address = createAddress(params.address)
		try {
			if (params.blockTag === 'pending') {
				const mineResult = await getPendingClient(client)
				if (mineResult.errors) {
					return maybeThrowOnFail(throwOnFail, {
						errors: mineResult.errors,
						address: params.address,
						balance: 0n,
						/**
						 * @type {`0x${string}`}
						 */
						storageRoot: '0x',
						nonce: 0n,
						/**
						 * @type {`0x${string}`}
						 */
						deployedBytecode: '0x',
						/**
						 * @type {`0x${string}`}
						 */
						codeHash: '0x',
						isContract: false,
						isEmpty: true,
					})
				}
				return getAccountHandler(mineResult.pendingClient, options)({ throwOnFail, ...params, blockTag: 'latest' })
			}
			if (params.blockTag !== 'latest' && params.blockTag !== undefined) {
				const block = await vm.blockchain.getBlockByTag(params.blockTag)
				const clonedVm = await cloneVmWithBlockTag(client, block)
				if (clonedVm instanceof ForkError || clonedVm instanceof InternalError) {
					return maybeThrowOnFail(throwOnFail, {
						errors: [clonedVm],
						address: params.address,
						balance: 0n,
						/**
						 * @type {`0x${string}`}
						 */
						storageRoot: '0x',
						nonce: 0n,
						/**
						 * @type {`0x${string}`}
						 */
						deployedBytecode: '0x',
						/**
						 * @type {`0x${string}`}
						 */
						codeHash: '0x',
						isContract: false,
						isEmpty: true,
					})
				}
				return getAccountHandler(
					{ ...client, getVm: () => Promise.resolve(clonedVm) },
					options,
				)({ throwOnFail, ...params, blockTag: 'latest' })
			}
			const res = await vm.stateManager.getAccount(address)
			if (!res) {
				return maybeThrowOnFail(throwOnFail, {
					address: params.address,
					balance: 0n,
					/**
					 * @type {`0x${string}`}
					 */
					storageRoot: '0x',
					nonce: 0n,
					/**
					 * @type {`0x${string}`}
					 */
					deployedBytecode: '0x',
					errors: [new AccountNotFoundError(`account ${params.address} not found`)],
					/**
					 * @type {`0x${string}`}
					 */
					codeHash: '0x',
					isContract: false,
					isEmpty: true,
				})
			}
			const code = res?.codeHash !== undefined ? bytesToHex(await vm.stateManager.getContractCode(address)) : '0x'

			return {
				// TODO some of these fields are not in the api and should be added to @tevm/actions
				address: params.address,
				balance: res.balance,
				codeHash: bytesToHex(res.codeHash),
				isContract: res.isContract(),
				isEmpty: res.isEmpty(),
				deployedBytecode: code,
				nonce: res.nonce,
				storageRoot: bytesToHex(res.storageRoot),
				...(params.returnStorage
					? {
							storage: Object.fromEntries(
								Object.entries(await vm.stateManager.dumpStorage(address)).map(([key, value]) => [
									`0x${key}`,
									/** @type {import('../common/Hex.js').Hex}*/ (value),
								]),
							),
						}
					: {}),
			}
		} catch (e) {
			let err = e
			// TODO this isn't clean
			// What we are doing here is checking if we threw a known error or not
			if (typeof e !== 'object' || e === null || !('_tag' in e)) {
				err = new InternalError('UnexpectedError in getAccountHandler', { cause: /** @type {any}*/ (e) })
			}
			errors.push(/** @type any*/ (err))
			return maybeThrowOnFail(throwOnFail, {
				errors,
				address: params.address,
				balance: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				storageRoot: '0x',
				/**
				 * @type {`0x${string}`}
				 */
				codeHash: '0x',
				nonce: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				deployedBytecode: '0x',
				isContract: false,
				isEmpty: true,
			})
		}
	}
