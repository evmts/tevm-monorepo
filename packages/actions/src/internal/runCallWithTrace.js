import { bytesToHex, numberToHex } from '@tevm/utils'
/**
 * @internal
 * Prepares a trace to be listened to
 * @param {import('@tevm/vm').TevmVm} vm
 * @param {import('@tevm/base-client').BaseClient['logger']} logger
 * @param {import('@tevm/evm').EvmRunCallOpts} params
 * @returns {Promise<import('@tevm/evm').EvmResult & {trace: import('@tevm/actions-types').DebugTraceCallResult}>}
 */
export const runCallWithTrace = async (vm, logger, params) => {
	/**
	 * As the evm runs we will be updating this trace object
	 * and then returning it
	 */
	const trace = {
		gas: 0n,
		/**
		 * @type {import('@tevm/utils').Hex}
		 */
		returnValue: '0x0',
		failed: false,
		/**
		 * @type {Array<import('@tevm/actions-types').DebugTraceCallResult['structLogs'][number]>}
		 */
		structLogs: [],
	}

	/**
	 * On every step push a struct log
	 */
	vm.evm.events?.on('step', async (step, next) => {
		logger.debug(step, 'runCallWithTrace: new evm step')
		trace.structLogs.push({
			pc: step.pc,
			op: step.opcode.name,
			gasCost: BigInt(step.opcode.fee) + (step.opcode.dynamicFee ?? 0n),
			gas: step.gasLeft,
			depth: step.depth,
			stack: step.stack.map((code) => numberToHex(code)),
		})
		next?.()
	})

	/**
	 * After any internal call push error if any
	 */
	vm.evm.events?.on('afterMessage', (data, next) => {
		logger.debug(data.execResult, 'runCallWithTrace: new message result')
		if (data.execResult.exceptionError !== undefined && trace.structLogs.length > 0) {
			// Mark last opcode trace as error if exception occurs
			const nextLog = trace.structLogs[trace.structLogs.length - 1]
			if (!nextLog) {
				throw new Error('No structLogs to mark as error')
			}
			// TODO fix this type
			Object.assign(nextLog, {
				error: data.execResult.exceptionError,
			})
		}
		next?.()
	})

	const runCallResult = await vm.evm.runCall(params)

	logger.debug(runCallResult, 'runCallWithTrace: evm run call complete')

	trace.gas = runCallResult.execResult.executionGasUsed
	trace.failed = runCallResult.execResult.exceptionError !== undefined
	trace.returnValue = bytesToHex(runCallResult.execResult.returnValue)

	return {
		...runCallResult,
		trace,
	}
}
