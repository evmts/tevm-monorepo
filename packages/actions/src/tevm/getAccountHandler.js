import { createError } from './createError.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { EthjsAddress } from '@tevm/utils'
import { bytesToHex, hexToBytes } from '@tevm/utils'
import { validateGetAccountParams } from '@tevm/zod'

/**
 * Creates an GetAccountHandler for handling account params with Ethereumjs VM
 * @param {Pick<import('@tevm/base-client').BaseClient, 'getVm'>} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('@tevm/actions-types').GetAccountHandler}
 */
export const getAccountHandler =
	(client, options = {}) =>
	async ({ throwOnFail = options.throwOnFail ?? true, ...params }) => {
		const vm = await client.getVm()
		/**
		 * @type {Array<import('@tevm/errors').GetAccountError>}
		 */
		const errors = validateGetAccountParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, {
				errors,
				address: params.address,
				balance: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				storageRoot: '0x',
				nonce: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				deployedBytecode: '0x',
				/**
				 * @type {`0x${string}`}
				 */
				codeHash: '0x',
				isContract: false,
				isEmpty: true,
			})
		}

		const address = new EthjsAddress(hexToBytes(params.address))
		try {
			const res = await vm.stateManager.getAccount(address)
			if (!res) {
				return maybeThrowOnFail(throwOnFail, {
					address: params.address,
					balance: 0n,
					/**
					 * @type {`0x${string}`}
					 */
					storageRoot: '0x',
					nonce: 0n,
					/**
					 * @type {`0x${string}`}
					 */
					deployedBytecode: '0x',
					errors: [
						createError(
							'AccountNotFoundError',
							`account ${params.address} not found`,
						),
					],
					/**
					 * @type {`0x${string}`}
					 */
					codeHash: '0x',
					isContract: false,
					isEmpty: true,
				})
			}
			const code =
				res?.codeHash !== undefined
					? bytesToHex(await vm.stateManager.getContractCode(address))
					: '0x'
			return {
				// TODO some of these fields are not in the api and should be added to @tevm/actions-types
				address: params.address,
				balance: res.balance,
				codeHash: bytesToHex(res.codeHash),
				isContract: res.isContract(),
				isEmpty: res.isEmpty(),
				deployedBytecode: code,
				nonce: res.nonce,
				storageRoot: bytesToHex(res.storageRoot),
			}
		} catch (e) {
			errors.push(
				createError(
					'UnexpectedError',
					typeof e === 'string'
						? e
						: e instanceof Error
						? e.message
						: 'unknown error',
				),
			)
			return maybeThrowOnFail(throwOnFail, {
				errors,
				address: params.address,
				balance: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				storageRoot: '0x',
				/**
				 * @type {`0x${string}`}
				 */
				codeHash: '0x',
				nonce: 0n,
				/**
				 * @type {`0x${string}`}
				 */
				deployedBytecode: '0x',
				isContract: false,
				isEmpty: true,
			})
		}
	}
