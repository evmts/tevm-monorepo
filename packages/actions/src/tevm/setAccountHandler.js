import { createError } from './createError.js'
import { maybeThrowOnFail } from './maybeThrowOnFail.js'
import { EthjsAccount, EthjsAddress } from '@tevm/utils'
import { hexToBytes, keccak256 } from '@tevm/utils'
import { validateSetAccountParams } from '@tevm/zod'

/**
 * Creates an SetAccountHandler for handling account params with Ethereumjs EVM
 * @param {Pick<import('@tevm/base-client').BaseClient, 'vm'>} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('@tevm/actions-types').SetAccountHandler}
 */
export const setAccountHandler = (client, options = {}) => async (params) => {
	const { throwOnFail = options.throwOnFail ?? true } = params
	/**
	 * @type {Array<import('@tevm/errors').SetAccountError>}
	 */
	const errors = validateSetAccountParams(params)
	if (errors.length > 0) {
		return maybeThrowOnFail(throwOnFail, { errors })
	}

	const address = new EthjsAddress(hexToBytes(params.address))
	try {
		await client.vm.stateManager.putAccount(
			address,
			new EthjsAccount(
				params.nonce,
				params.balance,
				params.storageRoot && hexToBytes(params.storageRoot),
				params.deployedBytecode &&
					hexToBytes(keccak256(params.deployedBytecode)),
			),
		)
		if (params.deployedBytecode) {
			await client.vm.stateManager.putContractCode(
				address,
				hexToBytes(params.deployedBytecode),
			)
		}
		await client.vm.stateManager.checkpoint()
		await client.vm.stateManager.commit()
		// TODO offer way of setting contract storage with evm.stateManager.putContractStorage
		return {}
	} catch (e) {
		errors.push(
			createError(
				'UnexpectedError',
				typeof e === 'string'
					? e
					: e instanceof Error
					? e.message
					: 'unknown error',
			),
		)
		return maybeThrowOnFail(throwOnFail, { errors })
	}
}
