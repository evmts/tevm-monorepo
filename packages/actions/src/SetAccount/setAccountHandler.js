import { createAddress } from '@tevm/address'
import { AccountNotFoundError, InternalError } from '@tevm/errors'
import { EthjsAccount } from '@tevm/utils'
import { hexToBytes, keccak256 } from '@tevm/utils'
import { getAccountHandler } from '../GetAccount/getAccountHandler.js'
import { maybeThrowOnFail } from '../internal/maybeThrowOnFail.js'
import { validateSetAccountParams } from './validateSetAccountParams.js'

/**
 * Creates an SetAccountHandler for handling account params with Ethereumjs EVM
 * @param {import("@tevm/node").TevmNode} client
 * @param {object} [options]
 * @param {boolean} [options.throwOnFail] whether to default to throwing or not when errors occur
 * @returns {import('./SetAccountHandlerType.js').SetAccountHandler}
 */
export const setAccountHandler =
	(client, options = {}) =>
	async (params) => {
		const { throwOnFail = options.throwOnFail ?? true } = params
		/**
		 * @type {Array<import('../SetAccount/TevmSetAccountError.js').TevmSetAccountError>}
		 */
		const errors = validateSetAccountParams(params)
		if (errors.length > 0) {
			return maybeThrowOnFail(throwOnFail, { errors })
		}

		const address = createAddress(params.address)

		/**
		 * @type {Array<Promise<any>>}
		 */
		const promises = []
		try {
			const vm = await client.getVm()

			// check if account exists
			const account = await getAccountHandler(client)({ ...params, throwOnFail: false })
			if (account.errors?.length && !(account.errors[0] instanceof AccountNotFoundError)) {
				client.logger.error('there was an unexpected error getting account', account.errors)
				throw account.errors.length > 1 ? new AggregateError(account.errors) : account.errors[0]
			}

			promises.push(
				vm.stateManager.putAccount(
					address,
					new EthjsAccount(
						params.nonce ?? account?.nonce,
						params.balance ?? account?.nonce,
						(params.storageRoot && hexToBytes(params.storageRoot)) ??
							(account?.storageRoot !== undefined && account?.storageRoot !== '0x'
								? hexToBytes(account.storageRoot)
								: undefined),
						(params.deployedBytecode && hexToBytes(keccak256(params.deployedBytecode))) ??
							(account?.deployedBytecode !== undefined ? hexToBytes(keccak256(account.deployedBytecode)) : undefined),
					),
				),
			)
			if (params.deployedBytecode) {
				promises.push(vm.stateManager.putContractCode(address, hexToBytes(params.deployedBytecode)))
			}
			// state clears state first stateDiff does not
			if (params.state) {
				await vm.stateManager.clearContractStorage(address)
			}
			const state = params.state ?? params.stateDiff
			if (state) {
				for (const [key, value] of Object.entries(state)) {
					promises.push(
						vm.stateManager.putContractStorage(
							address,
							hexToBytes(/** @type {import('@tevm/utils').Hex}*/ (key), { size: 32 }),
							hexToBytes(value),
						),
					)
				}
			}
			const results = await Promise.allSettled(promises)
			for (const result of results) {
				if (result.status === 'rejected') {
					errors.push(new InternalError('Unable to put storage', { cause: result.reason }))
				}
			}

			if (errors.length > 0) {
				return maybeThrowOnFail(throwOnFail, { errors })
			}
			await vm.stateManager.checkpoint()
			await vm.stateManager.commit(false)

			return {}
		} catch (e) {
			errors.push(new InternalError('Unexpected error setting account', { cause: e }))
			return maybeThrowOnFail(throwOnFail, { errors })
		}
	}
