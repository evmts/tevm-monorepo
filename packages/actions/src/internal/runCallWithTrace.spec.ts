import { createTevmNode } from '@tevm/node'
import { TestERC20 } from '@tevm/test-utils'
import { encodeFunctionData, hexToBytes } from '@tevm/utils'
import { EthjsAddress } from '@tevm/utils'
import { describe, expect, it } from 'vitest'
import { getAccountHandler } from '../GetAccount/getAccountHandler.js'
import { setAccountHandler } from '../SetAccount/setAccountHandler.js'
import { runCallWithTrace } from './runCallWithTrace.js'

const ERC20_ADDRESS = `0x${'1'.repeat(40)}` as const
const ERC20_BYTECODE = TestERC20.deployedBytecode
const ERC20_ABI = TestERC20.abi

describe('runCallWithTrace', () => {
	it('should execute a contract call with tracing', async () => {
		const client = createTevmNode()

		expect(
			(
				await setAccountHandler(client)({
					address: ERC20_ADDRESS,
					deployedBytecode: ERC20_BYTECODE,
				})
			).errors,
		).toBeUndefined()

		expect(
			await getAccountHandler(client)({
				address: ERC20_ADDRESS,
			}),
		).toMatchObject({
			address: ERC20_ADDRESS,
			deployedBytecode: ERC20_BYTECODE,
		})

		// Call the contract with trace
		const vm = await client.getVm().then((vm) => vm.deepCopy())
		const head = await vm.blockchain.getCanonicalHeadBlock()
		await vm.stateManager.setStateRoot(head.header.stateRoot)

		const params = {
			data: hexToBytes(
				encodeFunctionData({
					abi: ERC20_ABI,
					functionName: 'balanceOf',
					args: [ERC20_ADDRESS],
				}),
			),
			gasLimit: 16784800n,
			to: EthjsAddress.fromString(ERC20_ADDRESS),
			block: await vm.blockchain.getCanonicalHeadBlock(),
			origin: EthjsAddress.zero(),
			caller: EthjsAddress.zero(),
		}

		const result = await runCallWithTrace(vm, client.logger, params)

		expect(result).toHaveProperty('trace')
		expect(result.trace).toMatchObject({
			gas: expect.any(BigInt),
			returnValue: expect.any(String),
			failed: expect.any(Boolean),
			structLogs: expect.any(Array),
		})

		expect(result.execResult.returnValue).toMatchSnapshot()
		expect(result.createdAddress).toMatchSnapshot()
		expect(result.trace.gas).toMatchSnapshot()
		expect(result.trace.returnValue).toMatchSnapshot()
		expect(result.trace.structLogs).toMatchSnapshot()
	})
})
