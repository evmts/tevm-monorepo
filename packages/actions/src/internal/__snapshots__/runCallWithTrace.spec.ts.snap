// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`runCallWithTrace should execute a contract call with tracing 1`] = `
Uint8Array [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;

exports[`runCallWithTrace should execute a contract call with tracing 2`] = `undefined`;

exports[`runCallWithTrace should execute a contract call with tracing 3`] = `Any<BigInt>`;

exports[`runCallWithTrace should execute a contract call with tracing 4`] = `Any<String>`;

exports[`runCallWithTrace should execute a contract call with tracing 5`] = `Any<Array>`;
